### PR TITLE
Added cache bypass option

### PR DESCRIPTION
**Pull request by joosthooz**
PR to address #10 
Subtasks:
- [x] An option to bypass the caching mechanism for intermediate formats (preferably keeping disk usage to a minimum, so immediately removing files once not needed anymore) -> [d8216fe](https://github.com/conbench/datalogistik/pull/31/commits/d8216fe1c039f0822ddc802fa347ee9d69440f0f)
- [x] An option to remove cache entries by literal specifying the subpath to remove

### CODE REVIEW

File: .idea/modules.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (austin3dickey) [inline .idea/modules.xml:1]: Could you add this to the .gitignore please?

File: datalogistik/cli.py [hunk 2]
```diff
@@ -66,18 +83,37 @@ def parse_args():
...
         default=0,
         help="Partition the dataset using this maximum number of rows per file",
     )
-
+    gen_parser.add_argument(
+        "-b",
+        "--bypass-cache",
+        action="store_true",
+        help="Do not store any copies of the dataset in the cache",
+    )
```
Comment (austin3dickey) [inline datalogistik/cli.py:98]: Could you add all these new options to the README?

**Review by austin3dickey — CHANGES_REQUESTED** [review]
nice, looks good! are you going to do the third subtask in this PR?

**Comment by austin3dickey** [comment]
To fix the build, you'll have to fix the CI yml [here](https://github.com/conbench/datalogistik/blob/main/.github/workflows/actions.yml#L123) and [here](https://github.com/conbench/datalogistik/blob/main/.github/workflows/actions.yml#L148) with the new CLI structure

File: .idea/modules.xml (hunk 1 above)
Comment (joosthooz) [reply .idea/modules.xml]: Sorry, this should not have been committed! I just removed it, it seems weird to have such a tool-specific item in the .gitignore

File: datalogistik/cli.py (hunk 2 above)
Comment (joosthooz) [reply datalogistik/cli.py:98]: Added in [1abcacc](https://github.com/conbench/datalogistik/pull/31/commits/1abcacc132406bd485f5e2643e41110b9a7d482b)

**Comment by joosthooz** [comment]
> nice, looks good! are you going to do the third subtask in this PR?
No, I created a new issue for that: #33

File: datalogistik/cli.py [hunk 3]
```diff
@@ -66,18 +83,37 @@ def parse_args():
...
         "-p",
         "--partition-max-rows",
         type=int,
         default=0,
         help="Partition the dataset using this maximum number of rows per file",
     )
-
+    gen_parser.add_argument(
+        "-b",
+        "--bypass-cache",
```
Comment (austin3dickey) [inline datalogistik/cli.py:95]: oh could you please add this to the README too?

File: README.rst [hunk 4]
```diff
@@ -9,13 +9,20 @@ generate some datasets such as TPC-H by calling their (external) generator progr
 
 Usage::
 
-    datalogistik [-h] \
+    datalogistik generate <generate|cache>
```
Comment (austin3dickey) [inline README.rst:12]: Just a nitpick

```suggestion
    datalogistik <generate|cache>
```

File: datalogistik/util.py [hunk 5]
```diff
@@ -140,6 +136,77 @@ def clean_cache_dir(path):
...
+        if pathlib.Path(dirpath) == cache_root:
+            continue
+        if not dirnames:
+            # walk up the directory tree up to the root to find a metadatafile
+            walking_path = pathlib.Path(dirpath)
+            valid_metadata_file_found = False
+            while walking_path.parent != cache_root:
+                metadata_file = pathlib.Path(walking_path, config.metadata_filename)
+                if metadata_file.exists():
+                    metadata = json.load(open(metadata_file))
```
Comment (austin3dickey) [inline datalogistik/util.py:152]: Would you mind using a context manager here so the `metadata_file` gets closed again? I don't want the open file to cause any headaches later

Comment (austin3dickey) [reply datalogistik/util.py:152]: And same down below where `open()` is used again

File: datalogistik/util.py [hunk 6]
```diff
@@ -140,6 +136,77 @@ def clean_cache_dir(path):
...
+        log.error(msg)
+        raise RuntimeError(msg)
+
+    metadata_file = pathlib.Path(path, config.metadata_filename)
+    if metadata_file.exists() and json.load(open(metadata_file)).get("files"):
+        # This is a valid dataset, continue pruning
+        pass
+    else:
+        # Not a valid dataset, check if this path is a subdir of a valid dataset
+        walking_path = pathlib.Path(path).parent
```
Comment (austin3dickey) [inline datalogistik/util.py:182]: I wonder if this chunk of code would be better as another util function? It's complex enough that separating its concerns would probably be easier to understand, and then you could call the new function both here and above

**Review by austin3dickey — CHANGES_REQUESTED** [review]
nice, a few more things

File: datalogistik/util.py (hunk 6 above)
Comment (joosthooz) [reply datalogistik/util.py]: Done in [dc3d499](https://github.com/conbench/datalogistik/pull/31/commits/dc3d499e816f52ffe7cd5f62dd38ed77fe9d324b)

File: datalogistik/util.py (hunk 5 above)
Comment (joosthooz) [reply datalogistik/util.py]: Done in [da57570](https://github.com/conbench/datalogistik/pull/31/commits/da5757024f938e9d280475c99b578db8af2f037d)

File: README.rst (hunk 4 above)
Comment (joosthooz) [reply README.rst]: Done, thanks for catching

File: datalogistik/cli.py (hunk 3 above)
Comment (joosthooz) [reply datalogistik/cli.py:95]: Done in [ae4861e](https://github.com/conbench/datalogistik/pull/31/commits/ae4861e13f4beecd5f21eedaf320b2b7e90018fb)

**Comment by joosthooz** [comment]
Alright, I think I took care of the items, let me know what you think

**Review by austin3dickey — APPROVED** [review]
looks good!

**Pull request merged**